components/nodemanager-service/pgdb/nodes_integration_test.go

```diff
@@ -366,3 +368,291 @@ func (suite *NodesIntegrationSuite) TestProjectsAreRoundtrippedThroughNodeLifecy
+		{Key: "last_run_status", Values: []string{"PASSED"}},
+	})
+	if err != nil {
+		suite.FailNow(err.Error())
+	}
+	suite.Equal(1, len(listNodes))
+
+}
+
+func (suite *NodesIntegrationSuite) TestFilterByScanDataPenultStatus() {
+	nowTime := ptypes.TimestampNow()
+	node := &manager.NodeMetadata{
+		Uuid:            "1223-4254-2424-1322",
+		Name:            "my really cool client run node",
+		PlatformName:    "debian",
+		PlatformRelease: "8.6",
+		LastContact:     nowTime,
+		SourceId:        "",
+		SourceRegion:    "",
+		SourceAccountId: "",
+		JobUuid:         "12343-232324-1231242",
+		ScanData: &nodes.LastContactData{
+			Id:      "1003-9254-2004-1322",
+			EndTime: nowTime,
+			Status:  nodes.LastContactData_FAILED,
+		},
+	}
+	err := suite.Database.ProcessIncomingNode(node)
+	if err != nil {
+		suite.FailNow(err.Error())
+	}
+
+	node.ScanData.Status = nodes.LastContactData_SKIPPED
+	err = suite.Database.ProcessIncomingNode(node)
+	if err != nil {
+		suite.FailNow(err.Error())
+	}
+
+	listNodes, _, err := suite.Database.GetNodes("", nodes.Query_ASC, 1, 100, []*common.Filter{
+		{Key: "last_scan_penultimate_status", Values: []string{"FAILED"}},
+	})
+	if err != nil {
+		suite.FailNow(err.Error())
+	}
+	suite.Equal(1, len(listNodes))
+	suite.Equal(nodes.LastContactData_FAILED, listNodes[0].GetScanData().GetPenultimateStatus())
+	suite.Equal(nodes.LastContactData_SKIPPED, listNodes[0].GetScanData().GetStatus())
+
+	listNodes, _, err = suite.Database.GetNodes("", nodes.Query_ASC, 1, 100, []*common.Filter{
+		{Key: "last_scan_penultimate_status", Values: []string{"FAILED"}},
+		{Key: "last_scan_status", Values: []string{"SKIPPED"}},
+	})
+	if err != nil {
+		suite.FailNow(err.Error())
+	}
+	suite.Equal(1, len(listNodes))
+}
+
+func (suite *NodesIntegrationSuite) TestFilterByScanDataAndRunDataStatus() {
+	nowTime := ptypes.TimestampNow()
+	node := &manager.NodeMetadata{
+		Uuid:            "1223-4254-2424-1322",
+		Name:            "my really cool client run node",
+		PlatformName:    "debian",
+		PlatformRelease: "8.6",
+		LastContact:     nowTime,
+		SourceId:        "",
+		SourceRegion:    "",
+		SourceAccountId: "",
+		JobUuid:         "12343-232324-1231242",
+		ScanData: &nodes.LastContactData{
+			Id:      "1003-9254-2004-1322",
+			EndTime: nowTime,
+			Status:  nodes.LastContactData_FAILED,
+		},
+	}
+	err := suite.Database.ProcessIncomingNode(node)
+	if err != nil {
+		suite.FailNow(err.Error())
+	}
+
+	node = &manager.NodeMetadata{
+		Uuid:            "1223-4254-2424-1322",
+		Name:            "my really cool client run node",
+		PlatformName:    "debian",
+		PlatformRelease: "8.6",
+		LastContact:     nowTime,
+		SourceId:        "",
+		SourceRegion:    "",
+		SourceAccountId: "",
+		JobUuid:         "12343-232324-1231242",
+		RunData: &nodes.LastContactData{
+			Id:      "1003-9254-2004-1322",
+			EndTime: nowTime,
+			Status:  nodes.LastContactData_PASSED,
+		},
+	}
```

Comment:
It might end up less neat than in my head, but with this test structure mostly repeated in every test, it might be nicer to extract the base metadata to a constant that gets copied and updated with whatever relevant values in each test. It'd make it clearer in each test what the significant values are.

Reply:
ya, that makes sense. there's a lot of stuff that's repeated